src/go/k8s/controllers/redpanda/redpanda_decommission_controller.go

```diff
@@ -0,0 +1,588 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/redpanda-data/redpanda/src/go/k8s/apis/redpanda/v1alpha1"
+)
+
+// +kubebuilder:rbac:groups="",namespace=default,resources=pods,verbs=get;list;watch;
+// +kubebuilder:rbac:groups="",namespace=default,resources=persistentvolumeclaims,verbs=get;list;update;patch;delete;watch
+// +kubebuilder:rbac:groups="",namespace=default,resources=persistentvolumes,verbs=get;list;update;patch;watch
+
+const (
+	RedpandaDecommissionLabelKey = "app.redpanda.io/decommission-start"
+)
+
+type DecommissionReconciler struct {
+	client.Client
+	// Namespace is where the pvc's are we are watching for
+	Namespace    string
+	OperatorMode bool
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DecommissionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// nolint:dupl // these are required event filters
+		For(&v1.StatefulSet{}).WithEventFilter(UpdateEventFilter).Complete(r)
+}
+
+func (r *DecommissionReconciler) Reconcile(c context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx, done := context.WithCancel(c)
+	defer done()
+
+	start := time.Now()
+	log := ctrl.LoggerFrom(ctx).WithName("DecommissionReconciler.Reconcile")
+
+	sts := &v1.StatefulSet{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, sts); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Examine if the object is under deletion
+	if !sts.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info(fmt.Sprintf("the statefulset %q is being deleted", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	// check if there exists a "decommission" annotation, if so process the decommission, otherwise
+	// check if required.
+	if sts.Annotations == nil {
+		sts.Annotations = make(map[string]string)
+	}
+	timestamp, decommissionStarted := sts.Annotations[RedpandaDecommissionLabelKey]
+
+	var err error
+	var result ctrl.Result
+
+	log.Info(fmt.Sprintf("is decomission started: %t", decommissionStarted))
+
+	// nolint:nestif // this is ok
+	if decommissionStarted {
+		timeStampParsed, errParsing := time.Parse(time.RFC3339, timestamp)
+		if errParsing != nil {
+			err = fmt.Errorf("could not parse timestamp %q: %w", timestamp, errParsing)
+			result = ctrl.Result{}
+		} else {
+			result, err = r.reconcileDecommission(ctx, timeStampParsed, sts)
+			if err != nil {
+				log.Info(fmt.Sprintf("found error trying to reconcile decommission, we will requeue: %s", err))
+			}
+		}
+	} else {
+		result, _, err = r.verifyIfNeedDecommission(ctx, sts)
+		if err != nil {
+			log.Info(fmt.Sprintf("found error trying to verify if we need to decommission: %s", err))
+		}
+	}
+
+	// Log reconciliation duration
+	durationMsg := fmt.Sprintf("reconciliation finished in %s", time.Since(start).String())
```

Comment:
```suggestion
	durationMsg := fmt.Sprintf("succesfull reconciliation finished in %s", time.Since(start).String())
```

the start timestamp is not used in error cases.